crates/holochain/src/conductor/p2p_agent_store.rs

```diff
@@ -28,6 +47,54 @@ pub async fn inject_agent_infos<'iter, I: IntoIterator<Item = &'iter AgentInfoSi
     Ok(p2p_put_all(&env, iter.into_iter()).await?)
 }
 
+/// Inject multiple agent info entries into the peer store in batches.
+pub async fn p2p_put_all_batch(env: EnvWrite, rx: tokio::sync::mpsc::Receiver<P2pBatch>) {
+    let stream = tokio_stream::wrappers::ReceiverStream::new(rx);
+    let mut stream = stream.ready_chunks(100);
+    while let Some(batch) = stream.next().await {
+        let mut responses = Vec::with_capacity(batch.len());
+        let (tx, rx) = tokio::sync::oneshot::channel();
+        let result = env
+            .async_commit(move |mut txn| {
+                'batch: for P2pBatch {
+                    peer_data: batch,
+                    result_sender: response,
+                } in batch
+                {
+                    for info in batch {
+                        match p2p_put_single(&mut txn, &info) {
+                            Ok(_) => (),
+                            Err(e) => {
+                                responses.push((Err(e), response));
+                                continue 'batch;
+                            }
+                        }
+                    }
+                    responses.push((Ok(()), response));
+                }
+                tx.send(responses)
+                    .expect("We never drop the receiver before this send");
+                DatabaseResult::Ok(())
```

Comment:
I don't understand the extra oneshot channel here, does `async_commit` not support returning `responses` directly?

Reply:
It's because the `async_commit` can error for example if the commit itself fails and then you won't get the responses back if they were returned in the result so I still need the list of responses so I can respond with an error if the entire commit fails.

Reply:
@freesig I think I would have rather seen you just open the transaction manually and commit manually, handling the errors the way you want, rather than doing this workaround. The point of `with_commit_sync` et al is to remove the boilerplate of the most common case, and if you have a case that doesn't fit that it's OK to just not use the convenience function. It's not boilerplate if it's not repeated many times.

Reply:
Yes but it's also about spawning the spawn_blocking thread (`async_commit` not `with_commit_sync`. That's a fair bit of code to duplicate and potentially get wrong / go out of date. I think it's cleaner to use the oneshot then duplicate both functions no?

Reply:
That's a fair point. It's a bit of a toss-up, but you also have a fair bit of code at the end of this function that wouldn't need to be there otherwise. I feel like the happy medium would be to write a more general version of async_commit which passes the conn to the closure rather than the transaction, and then you can have the async conn stuff encapsulated while still customizing the transaction handling, and async_commit can be written in terms of that more general function. But I'm fine with this too, I'm just using this as a chance to discuss style.